packages/cozy-stack-client/src/OAuthClient.js

```diff
@@ -0,0 +1,358 @@
+
+const defaultoauthOptions = {
+  clientID: '',
+  clientName: '',
+  clientKind: '',
+  clientSecret: '',
+  clientURI: '',
+  registrationAccessToken: '',
+  redirectURI: '',
+  softwareID: '',
+  softwareVersion: '',
+  logoURI: '',
+  policyURI: '',
+  notificationPlatform: '',
+  notificationDeviceToken: ''
+}
+
+export default class OAuthClient extends CozyStackClient {
+  constructor({ oauth, scope, ...options }) {
+    super(options)
+    this.oauthOptions = { ...defaultoauthOptions, ...oauth }
+    this.scope = scope
+  }
+
+  /**
+   * Checks if the client has his registration informations from the server
+   * @returns {boolean}
+   * @private
+   */
+  isRegistered() {
+    return this.oauthOptions.clientID !== ''
+  }
+
+  /**
+   * Converts a camel-cased data set to snake case, suitable for sending to the OAuth server
+   * @param   {object} data Initial data
+   * @returns {object} Formatted data
+   * @private
+   */
+  snakeCaseOAuthData(data) {
+    const mappedFields = {
+      softwareID: 'software_id',
+      softwareVersion: 'software_version',
+      clientID: 'client_id',
+      clientName: 'client_name',
+      clientKind: 'client_kind',
+      clientURI: 'client_uri',
+      logoURI: 'logo_uri',
+      policyURI: 'policy_uri',
+      notificationPlatform: 'notification_platform',
+      notificationDeviceToken: 'notification_device_token',
+      redirectURI: 'redirect_uris'
+    }
+
+    const result = {}
+
+    Object.keys(data).forEach(fieldName => {
+      const key = mappedFields[fieldName] || fieldName
+      const value = data[fieldName]
+      result[key] = value
+    })
+
+    // special case: turn redirect_uris into an array
+    if (
+      result['redirect_uris'] &&
+      result['redirect_uris'] instanceof Array === false
+    )
+      result['redirect_uris'] = [result['redirect_uris']]
+
+    return result
+  }
+
+  /**
+   * Converts a snake-cased data set to camel case, suitable for internal use
+   * @param   {object} data Initial data
+   * @returns {object} Formatted data
+   * @private
+   */
+  camelCaseOAuthData(data) {
+    const mappedFields = {
+      client_id: 'clientID',
+      client_name: 'clientName',
+      client_secret: 'clientSecret',
+      registration_access_token: 'registrationAccessToken',
+      software_id: 'softwareID'
+    }
+
+    const result = {}
+
+    Object.keys(data).forEach(fieldName => {
+      const key = mappedFields[fieldName] || fieldName
+      const value = data[fieldName]
+      result[key] = value
+    })
+
+    return result
+  }
+
+  /**
+   * Registers the currenly configured client with the OAuth server. 
+   * @throws {Error} When the client is already registered
+   * @returns {promise} A promise that resolves with a complete list of client information, including client ID and client secret.
+   */
+  async register() {
+    if (this.isRegistered()) throw new Error('Client already registered')
+
+    const data = await this.fetch(
```

Comment:
You don't catch fetch exception here.

Reply:
Uhm no, why do you think we should? I think it's desirable that an exception gets thrown if the `fetch` fails.

Reply:
Yes ok, it will reject the promise with the exception. But still, that's what we want, isn't it?